packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {EventEmitter2} from '@nestjs/event-emitter';
+import {NETWORK_FAMILY} from '@subql/common';
+import fetch from 'cross-fetch';
+import {NodeConfig} from '../../configure';
+import {getLogger} from '../../logger';
+import {BlockHeightMap} from '../../utils/blockHeightMap';
+import {DictionaryResponse, DictionaryVersion, IBlock, IDictionary, IDictionaryCtrl} from './types';
+import {subqlFilterBlocksCapabilities} from './v2';
+import {DictionaryV2Metadata} from './';
+
+export async function inspectDictionaryVersion(endpoint: string): Promise<DictionaryVersion> {
+  let resp: DictionaryV2Metadata;
+  try {
+    resp = await subqlFilterBlocksCapabilities(endpoint);
+    if (resp.supported.includes('complete')) {
+      return DictionaryVersion.v2Complete;
+    } else {
+      return DictionaryVersion.v2Basic;
+    }
+  } catch (e) {
+    logger.warn(`${e}. Try to use dictionary V1`);
+    return DictionaryVersion.v1;
+  }
+}
+
+const logger = getLogger('DictionaryService');
+export abstract class DictionaryService<DS, FB, D extends IDictionary<DS, FB>> implements IDictionaryCtrl<DS, FB> {
+  protected _dictionaries: D[] = [];
+
+  protected _currentDictionaryIndex: number | undefined;
+  constructor(
+    protected chainId: string,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly eventEmitter: EventEmitter2
+  ) {}
+
+  abstract initDictionaries(): Promise<void>;
+
+  init(dictionaries: D[]): void {
+    this._dictionaries = dictionaries;
+  }
+
+  get startHeight(): number {
+    return this.dictionary.startHeight;
+  }
+
+  get dictionary(): D {
+    if (this._dictionaries.length === 0) {
+      throw new Error(`No dictionaries available to use`);
+    }
+    if (this._currentDictionaryIndex === undefined || this._currentDictionaryIndex < 0) {
+      throw new Error(`Dictionary index is not set`);
+    }
+    return this._dictionaries[this._currentDictionaryIndex];
+  }
+
+  get useDictionary(): boolean {
+    if (
+      (!!this.nodeConfig.networkDictionaries || !!this.nodeConfig.dictionaryResolver) &&
```

Comment:
This shouldn't use nodeConfig.networkDictionaries. It breaks when using resolved dictionaries or manifest dictionaries